explainer/fenced_frame_config.md

```diff
@@ -0,0 +1,35 @@
+# Fenced frame configs
+
+
+# Introduction
+
+For use cases involving APIs that access cross-site data, we need to be able to load a fenced frame with content determined by the API without revealing information about the content to the embedding context. For example, with interest-based ads in [FLEDGE](https://github.com/WICG/turtledove), the winning ad that's returned from the auction auction depends on the user's cross-site interest group data, which we don't want to expose to the site that calls the auction. This document proposes a web-platform way of loading content into a fenced frame using an opaque object.
+
+
+# Proposed solution
+
+`<fencedframe>` has an attribute `config`, rather than `src`. APIs like FLEDGE return a WebIDL object called `FencedFrameConfig`, which has a series of fields that specify the behavior desired by the API (e.g. the ad url, width and height as seen from within the fenced frame, etc.). When the embedder stores this object into the `config` attribute, the fenced frame loads the context accordingly.
+  
+In order to hide information as described above, the browser _redacts_ `FencedFrameConfig` before sending it to the embedder. This means that certain fields which are sensitive, like the ad url, are replaced with an enum value `opaque`. The embedder may see whether there is a value defined for that field, but not what the value is. Likewise, when the embedder requests that a config be loaded into the fenced frame, the browser is responsible for looking up the config in a data structure in order to access the unredacted information.
+
+  
+## Turtledove Example
+
+When the SSP JS invokes the Turtledove API to run the ad auction, it gets back the `FencedFrameConfig` as the result, which is then used for rendering the fenced frame. This `FencedFrameConfig` has an opaque `src`, which maps to an actual ad url which is part of the interest group.
+
+
+```
+navigator.runAdAuction(myAuctionConfig).then((auctionWinnerConfig) => {
+  // auctionWinnerConfig value e.g.
+  // FencedFrameConfig {
+  //   'src': opaque ('ad.com/foo' internally)
+  //   ...
+  // }
+  var adFrame = document.createElement('fencedframe');
+  adFrame.setAttribute(config,auctionWinnerConfig);
```

Comment:
Two issues here:
  1. You have a comma accessor instead of a period
  2. `setAttribute()` won't actually work here since the `config` attribute is not a content attribute (string), but _only_ a WebIDL attribute, which are only settable via JS. You can tell this because we don't have `[Reflect]` in our WebIDL: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/fenced_frame/html_fenced_frame_element.idl;l=10;drc=d99163d64d43b85a9efcf0110a0cfa81886b4616

Reply:
The FF Opaque Source [explainer](https://github.com/WICG/fenced-frame/blob/master/explainer/opaque_src.md) example shows the auction result promise resolving to an opaque url...

```
navigator.runAdAuction(myAuctionConfig).then((auctionWinnerUrl) => {
  // auctionWinnerUrl value e.g.
  // urn:uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6
  var adFrame = document.createElement('fencedframe');
  adFrame.setAttribute(src,auctionWinnerUrl);
});
```

One chooses this path 'upstream' by returning a simple `'render': url` from `generateBid()`.


And the means to opt into the new way is to return 

`'render': {url: renderUrl, size: {width: renderWidth, height: renderHeight}},`

and then use 

```
navigator.runAdAuction(myAuctionConfig).then((auctionWinnerConfig) => {
  const adFrame = document.createElement('fencedframe');
  adFrame.config = auctionWinnerConfig;
});
```

Is that correct?

Reply:
> and then use
> `navigator.runAdAuction(myAuctionConfig).then((auctionWinnerConfig) => { [.....]`

Actually the way you'll get FLEDGE's promise to resolve a `FencedFrameConfig` (instead of a URN) is by having `resolveToConfig: true` _inside_ the `myAuctionConfig` (that you pass into `runAdAuction()`). That's how you communicate your preference to deal in terms of configs (instead of URNs) to FLEDGE. This behavior isn't enabled yet though, so if you do it today you'll still get a URN. In fact, even once the behavior is enabled, it will only be rolled out gradually to more users for testing, so if you pass in `resolveToConfig` there will be a transition period where FLEDGE _may_ still give you a URN. Long-term, FLEDGE will always resolve to a `FencedFrameConfig` with or without `resolveToConfig: true` in the auction configuration.

The reason all of this is not super clear yet is because I have not made the PR to the FLEDGE repository which describes exactly how to take advantage of this, but I'll be doing that soon.

----

Regarding your first question:

> And the means to opt into the new way is to return
> `'render': {url: renderUrl, size: {width: renderWidth, height: renderHeight}},`

I am not sure about this (not as familiar with FLEDGE). @gtanzer do you know?

Reply:
Yes, the combination of those things is correct. In order to hook into the new size behavior in FLEDGE, you will need to  declare ad sizes in the interest group declaration, make a bid that includes both url and size, put `resolveToConfig` in the auction config, and then store the winning config in `.config` as @dmdabbs's first example shows. (You can't/don't need to modify a config that gets returned from FLEDGE; it already has everything baked in.)

Reply:
Actually, it may not be necessary to use configs to hook into the FLEDGE size changes: we would just need to do some extra implementation work to make urns work. We'll get back to you on that.

Reply:
An IG ad's size metdata may be [percentages](https://github.com/WICG/turtledove/pull/417/commits/d63d8e13537e697c93c52f7929da3cc06cb9ce4e?short_path=d65ba97#diff-d65ba9778fe3af46de3edfce2266b5b035192f8869280ec07179963b81f4e624), 

>...screen dimension coordinates (e.g. 100sw or 100sh)

The penny hasn't dropped for me how the internal size (I forget your terminology) is set and the external (page-visible dimensions). 

Also how pubs/sellers deal with the URN->Newfangled Config rollout/transition.

Reply:
@domfarolino Re the original comment, updated to `adFrame.config = auctionWinnerConfig;`